sdks/java/core/src/main/java/org/apache/beam/sdk/testing/BigqueryMatcher.java

```diff
@@ -144,23 +144,30 @@ Bigquery newBigqueryClient(String applicationName) {
         .build();
   }
 
+  @Nonnull
   @VisibleForTesting
   QueryResponse queryWithRetries(Bigquery bigqueryClient, QueryRequest queryContent,
                                  Sleeper sleeper, BackOff backOff)
       throws IOException, InterruptedException {
+    QueryResponse response;
     IOException lastException = null;
     do {
       try {
-        return bigqueryClient.jobs().query(projectId, queryContent).execute();
+        response = bigqueryClient.jobs().query(projectId, queryContent).execute();
+        if (response != null) {
+          return response;
+        }
       } catch (IOException e) {
         // ignore and retry
         LOG.warn("Ignore the error and retry the query.");
         lastException = e;
       }
     } while(BackOffUtils.next(sleeper, backOff));
-    throw new IOException(
+
+    throw new RuntimeException(
         String.format(
-            "Unable to get BigQuery response after retrying %d times", MAX_QUERY_RETRIES),
+            "Unable to get BigQuery response after retrying %d times. Got response: null.",
```

Comment:
"Got response: null." is not accurate. However, you can remove it if you create a IOException yourself above.